[BREAKING] permanently set errorTypeForLoose from Warning to SyntaxError

christianparpart opened the pull request:
Closes #2916. Closes #3936. Closes #3937. Closes #3938.

Depends on #4545.

It seems that we do not have a test for it, at least there was nothing
that needed adaption.

### TODO:

- [x] EndToEndTests adaptions
- [x] documentation changes
- [x] adapting compilation tests

Commit:
Disallow loos assembly in Solidity by permanently setting it to SyntaxError (from Warning)

chriseth commented:
Please add a test.

christianparpart commented:
@chriseth so it turns out, that our v0.5.0 checklist (#4401) contains the errorTypeForLoose twice, and the first one is just a no-op, as `ReferenceResolver::visit(InlineAssembly const&)`'s invokation of `AsmAnalyzer` is not reporting anything (but only used to resolve references).

At the later stage (in TypeChecker), the change actually takes affect, and for that we have tests of course already.

I've adapted the tests accordingly.

Commit:
Update tests to remove support for loose assembly

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -277,7 +277,7 @@ bool ReferencesResolver::visit(InlineAssembly const& _inlineAssembly)
 	// We use the latest EVM version because we will re-run it anyway.
 	assembly::AsmAnalysisInfo analysisInfo;
-	boost::optional<Error::Type> errorTypeForLoose = m_experimental050Mode ? Error::Type::SyntaxError : Error::Type::Warning;
+	boost::optional<Error::Type> errorTypeForLoose = Error::Type::SyntaxError;
```
Either keep this here and in `TypeChecker` or merge into the `AsmAnalyzer` line in both cases. While first I was going to suggest not to keep the variable, at least it helps documenting the code.

Well, actually, as a second PR the parameter to `AsmAnalyzer` could be removed entirely.

chriseth replied:
@axic I thought we still might want to keep loose style for the `solc --assemble` mode, but we can also fully remove it.

christianparpart replied:
@axic I'd even go that far in removing that ctor param alltogether (no need for a seperate PR, IMHO), what do you think? /cc @chriseth

axic replied:
> @axic I thought we still might want to keep loose style for the solc --assemble mode, but we can also fully remove it.

Oh right, for that reason I would not change it too much in this PR then. If we decode to remove that, should be in a separate PR.

Commit:
Remove loose assembly features from the documentation

axic commented on docs/assembly.rst:
```diff
@@ -416,57 +414,7 @@ changes during the call, and thus references to local variables will be wrong.
-            jump(one)
-        three:
-    }
+Support for labels have been removed.  Please use functions, loops, if or switch statements instead.
```
Perhaps state that it was removed with 0.5.0?

chriseth commented on test/compilationTests/milestonetracker/RLP.sol:
```diff
@@ -381,19 +381,16 @@ library RLP {
                  let wOffset := add(tgt, 0x20)
-             tag_loop:
-                 jumpi(end, eq(i, words))
+
```
Could you try to merge that into https://github.com/androlo/standard-contracts/blob/master/contracts/src/codec/RLP.sol ?

christianparpart replied:
shure, I'd love to.

christianparpart replied:
for reference: https://github.com/androlo/standard-contracts/pull/8

ekpyron commented on test/compilationTests/stringutils/strings.sol:
```diff
@@ -506,15 +519,34 @@ library strings {
-                    ptr := add(ptr, needlelen)
-                    exit:
+
+                    if iszero(eq(and(mload(ptr), mask), needledata)) {
```
For the record: I think we need to have a close look at this one - I'm not sure whether the current code does what it should.

christianparpart replied:
yeah, I am aware of that and will look into this too. I'd highly appreciate others looking into it, too. :-)

axic commented on Changelog.md:
```diff
@@ -49,6 +49,7 @@ Breaking Changes:
  * Type Checker: Fallback function must be external. This was already the case in the experimental 0.5.0 mode.
  * Type Checker: Interface functions must be declared external. This was already the case in the experimental 0.5.0 mode.
  * Type Checker: Address members are not included in contract types anymore. An explicit conversion is now required before invoking an ``address`` member from a contract.
+ * Type Checker: Inline-assembly's loose error type raised from ``Warning`` to ``SyntaxError``.
```
I think this is not clear for users to understand. Something like `Disallow "loose assembly" syntax entirely. This means that jump, jump labels and non-functional instructions cannot be used anymore.`

axic commented on docs/assembly.rst:
```diff
@@ -416,57 +414,7 @@ changes during the call, and thus references to local variables will be wrong.
-            jump(one)
-        three:
-    }
+Support for labels has been removed. Please use functions, loops, if or switch statements instead.
```
Should state it was removed in 0.5.0.

axic commented:
@christianparpart can you please split the peephole optimiser changes into a separate PR?

axic commented:
Rebased.

axic commented on test/libsolidity/syntaxTests/viewPureChecker/assembly_jump_view_fail.sol:
```diff
@@ -4,5 +4,4 @@ contract C {
     }
```
I think these two jump tests in `viewPureChecker` are not needed anymore.

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -390,11 +390,11 @@ BOOST_AUTO_TEST_CASE(unsatisfied_version)
 		{Error::Type::Warning, "Variable is shadowed in inline assembly by an instruction of the same name"},
-		{Error::Type::Warning, "The use of non-functional instructions is deprecated."},
+		{Error::Type::SyntaxError, "Top-level expressions are not supposed to return values"},
 		{Error::Type::DeclarationError, "Unbalanced stack"}
```
Probably `pop(returndatasize())` would be better to avoid these two errors.

christianparpart replied:
I agree, however, since the error was there before already I didn't feel quite sure of whether to `pop` it or not. Thanks for clarifying.

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -413,7 +413,7 @@ BOOST_AUTO_TEST_CASE(create2_as_variable)
 		{Error::Type::Warning, "The \"create2\" instruction is not supported by the VM version"},
 		{Error::Type::DeclarationError, "Unbalanced stack"},
-		{Error::Type::Warning, "not supposed to return values"}
+		{Error::Type::SyntaxError, "Top-level expressions are not supposed to return values (this expression returns 1 value). Use ``pop()`` or assign them."}
```
Same here with `pop(create2(...))`.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11255,21 +11209,6 @@ BOOST_AUTO_TEST_CASE(recursive_structs)
-		contract C {
-			function f() public {
-				assembly {
-					invalid
```
This can be changed into `invalid()` and kept.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -8910,52 +8910,6 @@ BOOST_AUTO_TEST_CASE(inline_assembly_storage_access_via_pointer)
-				assembly {
-					_x
-					jump(g)
-					pop
```
This should be added to the pure assembly tests (`test/InlineAssembly`).

christianparpart replied:
@axic I removed those two tests because it is not possible anymore to do jumps or function calls to an outside function. What could I move to `test/InlineAssembly`, or, what exactly should I test for?

Thanks,
Christian.

axic replied:
It is not possible within inline assembly, but still possible with the pure assembler, IIRC.

axic replied:
Sorry you are right, didn't notice for a second that `f` is a Solidity function.

Commit:
Drop swap_peephole_optimization test from EndToEndTests

Already covered by libevmasm/Optimiser: peephole_noncommutative_swap1, peephole_commutative_swap1 and peephole_swap_comparison

Commit:
Drops tests for inline assembly jumps and function access as both isn't possible anymore.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11255,21 +11224,6 @@ BOOST_AUTO_TEST_CASE(recursive_structs)
 	ABI_CHECK(callContractFunction("f()"), encodeArgs(u256(1)));
 }
 
-BOOST_AUTO_TEST_CASE(invalid_instruction)
```
Can you keep this at the original location in the file?
